builder/upcloud/config.go

```diff
@@ -51,24 +52,53 @@ type Config struct {
 	common.PackerConfig `mapstructure:",squash"`
 	Comm                communicator.Config `mapstructure:",squash"`
 
-	// Required configuration values
-	Username    string `mapstructure:"username"`
-	Password    string `mapstructure:"password"`
-	Zone        string `mapstructure:"zone"`
-	StorageUUID string `mapstructure:"storage_uuid"`
-	StorageName string `mapstructure:"storage_name"`
+	// The username to use when interfacing with the UpCloud API.
+	Username string `mapstructure:"username" required:"true"`
 
-	// Optional configuration values
-	TemplatePrefix string        `mapstructure:"template_prefix"`
-	TemplateName   string        `mapstructure:"template_name"`
-	StorageSize    int           `mapstructure:"storage_size"`
-	Timeout        time.Duration `mapstructure:"state_timeout_duration"`
-	CloneZones     []string      `mapstructure:"clone_zones"`
+	// The password to use when interfacing with the UpCloud API.
+	Password string `mapstructure:"password" required:"true"`
 
+	// The zone in which the server and template should be created (e.g. nl-ams1).
+	Zone string `mapstructure:"zone" required:"true"`
+
+	// The UUID of the storage you want to use as a template when creating the server.
+	StorageUUID string `mapstructure:"storage_uuid" required:"true"`
+
+	// The name of the storage that will be used to find the first matching storage in the list of existing templates.
+	//
+	// Note that `storage_uuid` parameter has higher priority. You should use either `storage_uuid` or `storage_name` for not strict matching (e.g "ubuntu server 20.04").
+	StorageName string `mapstructure:"storage_name" required:"true"`
```

Comment:
Is it actually required? If I understand correctly, we can use either uuid or name?

Reply:
Yep this is (for me at least) little tricky. I think that there is no way to describe that one of those needs to be defined so I decided to mark them both as required. Would it be more clear if we set `storage_name` as not required? Not 100% sure but I think that these `required` tags only effect docs. I think that validator still accepts config if there is no `storage_name` or `storage_uuid`

Reply:
Yeah, I thought it might be the case.
Technically speaking, they are both optional, but marking them as such does not help the user much :P
If it were up to me I would go with second option (so `storage_name` not required, `storage_uuid` required + some more elaborate comments that `storage_name` can actually be used _instead of_ `storage_uuid`), as I assume most users are going to rely on uuids anyway. But I might be wrong here.

Anyway, this is minor detail, not a blocker really especially since the added comments at the bottom make the situation fairly clear :)

Reply:
I would probably use storage name to match storage so that config would be easier to read :)
Anyway you have a good  point and at this point it's probably safer to mark that optional because it has been optional before.
I removed that required flag and added note also under `storage_uuid`